### PR TITLE
[WIP] Backport C++17 type_traits and C++20 chrono

### PR DESCRIPTION
**Pull request by trxcllnt**
PR to apply the backports we did for C++17 and 20 features for cuDF.

Related cuDF PR: https://github.com/rapidsai/cudf/pull/6275

### CODE REVIEW

File: libcxx/include/chrono [hunk 1]
```diff
@@ -1061,7 +1061,7 @@ private:
     rep __rep_;
 public:
 
-#if !defined(_LIBCUDACXX_CXX03_LANG)
+#if !defined(_LIBCUDACXX_CXX03_LANG) && !defined(__NVCC__)
```
Comment (trxcllnt) [inline libcxx/include/chrono:1064]: This line is suspect, but required in order to declare `__shared__` durations in kernels with nvcc <= 10.2.

Comment (griwes) [reply libcxx/include/chrono:1064]: Please explain? This is counterintuitive to me, did it not understand `= default` as trivial when appropriate?

Anyway if anything, to enable this, this should be a version check and not just blank disable of `= default` for all versions of NVCC.

File: libcxx/include/chrono [hunk 2]
```diff
@@ -2758,6 +2758,7 @@ inline constexpr year_month_weekday_last& year_month_weekday_last::operator-=(co
 inline constexpr year_month_weekday_last& year_month_weekday_last::operator+=(const years& __dy)  noexcept { *this = *this + __dy; return *this; }
 inline constexpr year_month_weekday_last& year_month_weekday_last::operator-=(const years& __dy)  noexcept { *this = *this - __dy; return *this; }
 
+#if _LIBCUDACXX_STD_VER > 17
```
Comment (griwes) [inline libcxx/include/chrono:2761]: What's C++20-specific about `hh_mm_ss`?

**Review by griwes — COMMENTED** [review]
In general this looks good to me, but I think we'll need to add more tests - we haven't really done any calendar testing to the libcu++ test suite, because C++20 wasn't a consideration yet - but if we backport it, we'll need to do that.

@jrhemstad this may be a good project for you to get some practice with the testing system on ;)

File: libcxx/include/chrono (hunk 2 above)
Comment (trxcllnt) [reply libcxx/include/chrono:2761]: `hh_mm_ss` doesn't compile on gcc-7, possibly only on ubuntu16.04 (haven't tested gcc-7 on ubuntu18.04).

File: libcxx/include/chrono (hunk 1 above)
Comment (trxcllnt) [reply libcxx/include/chrono:1064]: Here's an example of a line that errors (with CUDA 10.2, haven't tried 11 or < 10.2):

[`  __shared__ T temp_data[block_size];`](https://github.com/rapidsai/cudf/blob/c39e9168c6d08236f36e5c3203b3ba8d145e86ed/cpp/include/cudf/detail/copy_if.cuh#L108)

```
../../../../include/cudf/detail/copy_if.cuh (108): error: initializer not allowed for __shared__ variable 
```

**Comment by jrhemstad** [comment]
Working on porting remaining `<chrono>` tests from libc++ to libcu++ so we can make sure these changes are tested properly.

**Comment by jrhemstad** [comment]
This PR is superseded by https://github.com/NVIDIA/libcudacxx/pull/44 and https://github.com/NVIDIA/libcudacxx/pull/45

**Pull request closed without merging**